cibuildwheel/macos.py

```diff
@@ -21,12 +31,76 @@ def get_python_configurations(build_selector):
+
+    # Python bin folders on Mac don't symlink `python3` to `python`, and neither
+    # does PyPy for `pypy` or `pypy3`, so we do that so `python` and `pip` always
+    # point to the active configuration.
+    if os.path.exists(dest_dir):
+        shutil.rmtree(dest_dir)
+    os.makedirs(dest_dir)
+
+    os.symlink(os.path.join(installation_bin_path, python_executable), os.path.join(dest_dir, 'python'))
+    os.symlink(os.path.join(installation_bin_path, python_executable + '-config'), os.path.join(dest_dir, 'python-config'))
+    os.symlink(os.path.join(installation_bin_path, pip_executable), os.path.join(dest_dir, 'pip'))
+
+
+def install_cpython(version, url, symlinks_dir):
+    installed_system_packages = subprocess.check_output(['pkgutil', '--pkgs'], universal_newlines=True).splitlines()
+
+    # if this version of python isn't installed, get it from python.org and install
+    python_package_identifier = 'org.python.Python.PythonFramework-{}'.format(version)
+    if python_package_identifier not in installed_system_packages:
+        # download the pkg
+        download(url, '/tmp/Python.pkg')
+        # install
+        call(['sudo', 'installer', '-pkg', '/tmp/Python.pkg', '-target', '/'])
+        # patch open ssl
+        if version == '3.5':
+            open_ssl_patch_url = 'https://github.com/mayeut/patch-macos-python-openssl/releases/download/v1.0.2t/patch-macos-python-%s-openssl-v1.0.2t.tar.gz' % version
+            download(open_ssl_patch_url, '/tmp/python-patch.tar.gz')
+            call(['sudo', 'tar', '-C', '/Library/Frameworks/Python.framework/Versions/{}/'.format(version), '-xmf', '/tmp/python-patch.tar.gz'])
+
+    installation_bin_path = '/Library/Frameworks/Python.framework/Versions/{}/bin'.format(version)
+    python_executable = 'python3' if version[0] == '3' else 'python'
+    pip_executable = 'pip3' if version[0] == '3' else 'pip'
+    make_symlinks(symlinks_dir, installation_bin_path, python_executable, pip_executable)
+
+    return installation_bin_path
+
+
+def install_pypy(version, url, symlinks_dir):
+    pypy_tar_bz2 = url.rsplit('/', 1)[-1]
+    assert pypy_tar_bz2.endswith(".tar.bz2")
+    pypy_base_filename = os.path.splitext(os.path.splitext(pypy_tar_bz2)[0])[0]
+    installation_path = os.path.join('/tmp', pypy_base_filename)
+    if not os.path.exists(installation_path):
+        download(url, os.path.join("/tmp", pypy_tar_bz2))
+        call(['tar', '-C', '/tmp', '-xf', os.path.join("/tmp", pypy_tar_bz2)])
+
+    # fix PyPy 7.3.0 bug resulting in wrong macOS platform tag
+    if version.endswith("-v7.3.0") and version[0] == '3':
+        patch_file = os.path.abspath(os.path.join(os.path.dirname(__file__), 'resources', 'pypy3.6.patch'))
+        sysconfigdata_file = os.path.join(installation_path, 'lib_pypy', '_sysconfigdata.py')
+        subprocess.call(['patch', sysconfigdata_file, patch_file, '-N'])  # Always has nonzero return code
```

Comment:
Always non-zero - why?

patch should return 0 if '[all hunks are applied successfully](http://man7.org/linux/man-pages/man1/patch.1.html#DIAGNOSTICS)'... is that not the case here?

Reply:
Not always. Only first run end with 0. All others runs return non zero status. But in cibuildwheel run there is several use of cibuildwheel then after first run pypy installation is created. This can be changed by applying  this if only if condition from line 86 is true (ex. increase indent by 4 space of this code part.)

Reply:
I took this from @Czaki without thinking further about it. But if the first time it does return 0, I'll see if I can fix this. It's an extra check, anyway!

Reply:
@YannickJadoul Please see this https://github.com/Czaki/cibuildwheel/commit/09ddda3c6a888f4ea651ab929ad7a1326667021a

Reply:
Of course, that makes sense. Thanks, @Czaki! :-)